common/src/main/java/com/bakdata/quick/common/api/client/PartitionedMirrorClient.java

```diff
@@ -0,0 +1,135 @@
+    private final MirrorRequestManager requestManager;
+    private final Router<K> router;
+    private final List<MirrorHost> knownHosts;
+
+    /**
+     * Constructor that can be used when the mirror client is based on an IP or other non-standard host.
+     *
+     * @param topicName       the name of the topic
+     * @param mirrorHost      host to use
+     * @param client          http client
+     * @param keySerde        the serde for the key
+     * @param valueResolver   the value's {@link TypeResolver}
+     * @param partitionFinder strategy for finding partitions
+     */
+    public PartitionedMirrorClient(final String topicName, final MirrorHost mirrorHost, final HttpClient client,
+                                   final Serde<K> keySerde, final TypeResolver<V> valueResolver,
+                                   final PartitionFinder partitionFinder) {
+        this.streamsStateHost = StreamsStateHost.fromMirrorHost(mirrorHost);
+        this.client = client;
+        this.parser = new MirrorValueParser<>(valueResolver, client.objectMapper());
+        this.requestManager = new DefaultMirrorRequestManager(client);
+        log.info("Initializing partition router for the mirror at: {} and the topic: {}.",
+            this.streamsStateHost.getHost(), topicName);
+        final Map<Integer, String> response = this.makeRequestForPartitionHostMapping();
+        this.router = new PartitionRouter<>(keySerde, topicName, partitionFinder, response);
+        this.knownHosts = this.router.getAllHosts();
+    }
+
+    @Override
+    @Nullable
+    public V fetchValue(final K key) {
+        final MirrorHost currentKeyHost = router.findHost(key);
+        return this.requestManager.sendRequest(Objects.requireNonNull(currentKeyHost).forKey(key.toString()),
+            this.parser::deserialize);
+    }
+
+    @Override
+    public List<V> fetchAll() {
+        final List<V> valuesFromAllHosts = new ArrayList<>();
+        for (final MirrorHost host : this.knownHosts) {
+            final List<V> valuesFromSingleHost =
+                Objects.requireNonNullElse(this.requestManager.sendRequest(host.forAll(), this.parser::deserializeList),
+                    Collections.emptyList());
+            valuesFromAllHosts.addAll(valuesFromSingleHost);
+        }
+        return valuesFromAllHosts;
+    }
+
+    @Override
+    @Nullable
+    public List<V> fetchValues(final List<K> keys) {
+        return keys.stream().map(this::fetchValue).collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean exists(final K key) {
+        return this.fetchValue(key) != null;
+    }
+
+    private Map<Integer, String> makeRequestForPartitionHostMapping() {
+        final String url = this.streamsStateHost.getPartitionToHostUrl();
+        try (final ResponseBody responseBody = this.requestManager.makeRequest(url)) {
+            if (responseBody == null) {
+                throw new InternalErrorException("Response body was null.");
```

Comment:
This is still a MirrorClient. I would do it like the DefaultMirrorClient:
```suggestion
                throw new MirrorException("Response body was null.", HttpStatus.INTERNAL_SERVER_ERROR);
```